src/extension/app/utils/helix-admin.js

```diff
@@ -10,21 +10,130 @@
+  ONEDRIVE: 'Microsoft SharePoint',
+};
+
+/**
+ * Returns an error message from the server response.
+ * @param {Response} resp The response
+ * @returns {string} The error message or an empty string
+ */
+export function getServerError(resp) {
+  return resp?.headers?.get('x-error') || '';
+}
+
+/**
+ * Checks if the request has been rate-limited and returns the source.
+ * @param {Response} resp The response
+ * @returns {string} The rate limiter (see {@link RATE_LIMITER})
+ */
+function getRateLimiter(resp) {
+  if (resp.status === 429) {
+    return RATE_LIMITER.ADMIN;
+  }
+  const error = getServerError(resp);
+  if (resp.status === 503 && error.includes('(429)')) {
+    if (error.includes('onedrive')) {
+      return RATE_LIMITER.ONEDRIVE;
+    }
+  }
+  return RATE_LIMITER.NONE;
+}
+
+/**
+ * Returns the action name based on API and nature of request.
+ * @param {string} api The API endpoint called
+ * @param {boolean} [del] True if the request was destructive
+ */
+function getAction(api, del) {
+  let action = api;
+  if (api === 'preview' && del) {
+    action = 'delete';
+  }
+  if (api === 'live') {
+    action = del ? 'unpublish' : 'publish';
+  }
+  return action;
+}
+
+/**
+ * Shows a toast if the server returned an error.
+ * @param {AppStore} appStore The app store
+ * @param {string} action The action
+ * @param {Response} resp The response object
+ */
+function handleServerError(appStore, action, resp) {
+  if (resp.ok) {
+    return;
+  }
+
+  const limiter = getRateLimiter(resp);
+  if (limiter) {
+    appStore.showToast(appStore.i18n('error_429').replace('$1', limiter), 'warning');
+  }
+
+  let message = '';
+  if (action === 'status' && resp.status === 404) {
+    // special handling for status
+    message = appStore.i18n(appStore.isEditor()
+      ? 'error_status_404_document'
+      : 'error_status_404_content');
+  } else {
+    // error key fallbacks
+    message = appStore.i18n(`error_${action}_${resp.status}`)
+      || appStore.i18n(`error_${action}`)
+      || appStore.i18n('error_fatal');
+  }
+
+  appStore.showToast(
+    message.replace('$1', getServerError(resp)),
+    resp.status < 500 ? 'warning' : 'negative',
+    () => appStore.closeToast(),
+  );
+}
+
+/**
+ * Shows a toast if the request failed or did not contain valid JSON.
+ * @param {AppStore} appStore The app store
+ * @param {string} action The action
+ */
+function handleFatalError(appStore, action) {
+  // use standard error key fallbacks
+  const msg = appStore.i18n(`error_${action}_fatal`)
+      || appStore.i18n('error_fatal');
+  appStore.showToast(
+    msg.replace('$1', 'https://aemstatus.net/'),
+    'negative',
+    () => appStore.closeToast(),
+  );
+}
+
+/**
+ * Creates an Admin URL for an API and path.
  * @param {SiteStore} siteStore The sidekick config object
  * @param {string} api The API endpoint to call
  * @param {string} path The current path
  * @returns {URL} The admin URL
  */
 export function getAdminUrl({
```

Comment:
Could move this into the class and provide convenience methods for get login/logout url

Reply:
I'm sure it's still todo but there is a spot in siteStore that is calling the admin API.

Reply:
Thanks, will try to find all admin usages.